### PR TITLE
Refactor Tests

### PR DESCRIPTION
**Pull request by liampalmer43**
Started refactoring tests to Flix unit tests.
- Option tests refactored
- TODO: Rest of library tests

### CODE REVIEW

**Comment by liampalmer43** [comment]
Before I go further, I just want to confirm the goal is to do the unit test refactoring on all library tests (BigInt, Int8, Int16, ... etc.).

**Comment by magnus-madsen** [comment]
Yes, the goal is that there entire library will be tested by flixtest.

That *should* make things faster. I don't know why travis fails. I will have a look.

**Comment by magnus-madsen** [comment]
Your troubles might also be fixed by #350

**Comment by magnus-madsen** [comment]
What is the status on this one?

**Comment by liampalmer43** [comment]
I will keep adding refactored tests to this PR until all library tests are complete.  (This could be merged in though, and I could make a new PR for the remaining files.)

**Comment by magnus-madsen** [comment]
I merged map, so now I guess this has some conflicts.

**Comment by magnus-madsen** [comment]
The build issues should be fixed on master.

**Comment by magnus-madsen** [comment]
Hmm something seems wrong with this commit. The new changes in `LocalResource` seems to have been lost.

**Comment by liampalmer43** [comment]
I think that should fix it, `LocalResource` is the same here as it is on master now.  I think it just kept my old code locally when I merged in master.

**Comment by magnus-madsen** [comment]
Am I correct that the remaining cases are `TestInt16`, `TestInt32`, and `TestInt64` ?

**Comment by liampalmer43** [comment]
`TestInt32`, `TestInt64` and `TestList` are all that remain now.  So lets hold off on merging until I finish those.

File: main/test/ca/uwaterloo/flix/library/TestFloat32.flix [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// minPositiveValue                                                        //
+/////////////////////////////////////////////////////////////////////////////
+@test
+def minPositiveValue01: Bool = Float32.minPositiveValue() `assertEq!` (2.0f32 ** -149.0f32)
+
+/////////////////////////////////////////////////////////////////////////////
+// undefined                                                               //
+/////////////////////////////////////////////////////////////////////////////
+//@test
+//def undefined01: Bool = Float32.undefined() `assertEq!` (0.0f32 / 0.0f32)
```
Comment (liampalmer43) [inline main/test/ca/uwaterloo/flix/library/TestFloat32.flix]: This test fails for both Float32 and Float64, even (0.0f32 / 0.0f32) == (0.0f32 / 0.0f32) returns false, I'm not sure why.

File: main/test/ca/uwaterloo/flix/library/TestList.flix [hunk 2]
```diff
@@ -0,0 +1,1836 @@
...
+def foldRight03: Bool = List.foldRight((e, i) -> (i - e)*(e % 2 + 1), 100, 1 :: 2 :: Nil) `assertEq!` 194
+
+@test
+def foldRight04: Bool = List.foldRight((e, i) -> (i - e)*(e % 2 + 1), 100, 1 :: 2 :: 3 :: Nil) `assertEq!` 382
+
+/////////////////////////////////////////////////////////////////////////////
+// reduceOpt                                                               //
+/////////////////////////////////////////////////////////////////////////////
+//@test
+//def reduceOpt01: Bool = List.reduceOpt((a, b) -> a-b, Nil) `assertEq!` None
```
Comment (liampalmer43) [inline main/test/ca/uwaterloo/flix/library/TestList.flix]: I run into the typing issue with unit here:   `Can't apply Minus to type Unit near TestList.flix:992:50`
Maybe it should just be commented out for now?

**Comment by liampalmer43** [comment]
Hey @magnus-madsen, I think everything is good to go with this PR (there are two things I commented on above though).

File: main/test/ca/uwaterloo/flix/library/TestFloat32.flix (hunk 1 above)
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestFloat32.flix]: I will look into this.

File: main/test/ca/uwaterloo/flix/library/TestList.flix (hunk 2 above)
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestList.flix]: Yes, this is the bug I e-mailed about in #364. The fix is to add a type annotation somewhere. I will fix it.

File: main/test/ca/uwaterloo/flix/library/TestFloat32.flix (hunk 1 above)
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestFloat32.flix]: The issue is that (0.0f32 / 0.0f32) returns `NaN` and even though `undefined()` also returns `NaN` it is not the case that `NaN == NaN`.  This is discussed here #368. The solution will be to refactor the test to say:

```
def someName01: Bool = Float32.isNan(Float32.undefined())
// and
def someName02: Bool = Float32.isNan(0.0f32 / 0.0f32)

```

I have added a note to that issue.

**Comment by magnus-madsen** [comment]
@liampalmer43 Thanks for all your hard work in cleaning this up. More than 8000 lines changed (wow!). I really appreciate it. It means that we know have a real standard library that can be used for real applications allowing our language to be taken more seriously! :)

File: main/test/ca/uwaterloo/flix/library/TestList.flix (hunk 2 above)
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestList.flix]: I fixed the test cases in master. (The bug remains to be fixed).

**Pull request merged**